internal/pipeline/executor.go

```diff
@@ -281,3 +313,46 @@ func fnChain(pl *kptfilev1alpha2.Pipeline, pkgPath pkg.UniquePath) ([]kio.Filter
 	}
 	return runners, nil
 }
+
+// trackInputFiles records file paths of input resources in the hydration context.
+func trackInputFiles(hctx *hydrationContext, input []*yaml.RNode) error {
+	if hctx.inputFiles == nil {
+		hctx.inputFiles = sets.String{}
+	}
+	for _, r := range input {
+		path, _, err := kioutil.GetFileAnnotations(r)
+		if err != nil {
+			return fmt.Errorf("path annotation missing: %w", err)
+		}
+		hctx.inputFiles.Insert(path)
+	}
+	return nil
+}
+
+// trackOutputfiles records the file paths of output resources in the hydration
+// context. It should be invoked post hydration.
+func trackOutputFiles(hctx *hydrationContext) error {
+	outputSet := sets.String{}
+
+	for _, r := range hctx.root.resources {
+		path, _, err := kioutil.GetFileAnnotations(r)
+		if err != nil {
+			return fmt.Errorf("path annotation missing: %w", err)
+		}
+		outputSet.Insert(path)
+	}
+	hctx.outputFiles = outputSet
+	return nil
+}
+
+// pruneResources compares the input and output of the hydration and prunes
+// resources that are no longer present in the output of the hydration.
+func pruneResources(hctx *hydrationContext) error {
+	filesToBeDeleted := hctx.inputFiles.Difference(hctx.outputFiles)
+	for f := range filesToBeDeleted {
+		if err := os.Remove(filepath.Join(string(hctx.root.pkg.UniquePath), f)); err != nil {
```

Comment:
Same discussion as Donny's PR about semantics of path annotation.

Reply:
We realized we need to address this problem across the codebase (especially hydration that deals with path annotation much more). This code uses the `kioutl.GetPathAnnotation` abstraction to get the path annotation. If this abstraction returns OS-specific path (by doing the conversion from OS-agnostic to OS-specific...eventually), rest of the code will be using OS-specific code as today ('filepath.Join(..)`).

@Shell32-Natsu filed an issue https://github.com/kubernetes-sigs/kustomize/issues/3749 to track this in `kyaml` library.